File: v2/src/main/java/com/api/v2/people/domain/exposed/Person.java
package com.api.v2.people.domain.exposed;

import com.api.v2.people.enums.Gender;
import com.api.v2.people.requests.PersonRegistrationDto;
import org.springframework.data.annotation.Id;
import org.springframework.data.mongodb.core.mapping.Document;

import java.time.LocalDate;
import java.time.LocalDateTime;
import java.util.UUID;

@Document
public record Person (
    @Id
    String id,
    String firstName,
    String middleName,
    String lastName,
    String ssn,
    LocalDate birthDate,
    String phoneNumber,
    Gender gender,
    String email,
    LocalDateTime createdAt
) {

    public static Person of(PersonRegistrationDto registrationDto) {
        return new Person(
                UUID.randomUUID().toString(),
                registrationDto.firstName(),
                registrationDto.middleName(),
                registrationDto.lastName(),
                registrationDto.ssn(),
                registrationDto.birthDate(),
                registrationDto.phoneNumber(),
                registrationDto.gender(),
                registrationDto.email(),
                LocalDateTime.now()
        );
    }

}

File: v2/src/main/kotlin/com/api/v2/people/domain/exposed/Person.kt
package com.api.v2.people.domain.exposed

import com.api.v2.people.dtos.Address
import com.api.v2.people.enums.Gender
import com.api.v2.people.requests.PersonRegistrationDto
import com.api.v2.people.requests.PersonUpdatingDto
import org.springframework.data.annotation.Id
import org.springframework.data.mongodb.core.mapping.Document
import java.time.LocalDate
import java.time.LocalDateTime
import java.util.UUID

@Document
class Person private constructor(
    var firstName: String,
    var middleName: String?,
    var lastName: String,
    val sin: String,
    var birthDate: LocalDate,
    var gender: Gender,
    var address: Address,
    var email: String,
    var phoneNumber: String
) {

    @Id
    var id: String = UUID.randomUUID().toString()
    val createdAt: LocalDateTime = LocalDateTime.now()
    var updatedAt: LocalDateTime? = null

    companion object {
        fun of(registrationDto: PersonRegistrationDto): Person {
            return Person(
                registrationDto.firstName,
                registrationDto.middleName,
                registrationDto.lastName,
                registrationDto.sin,
                registrationDto.birthDate,
                registrationDto.gender,
                registrationDto.address,
                registrationDto.email,
                registrationDto.phoneNumber
            )
        }
    }

    fun update(updatingDto: PersonUpdatingDto) {
        firstName = updatingDto.firstName
        middleName = updatingDto.middleName
        lastName = updatingDto.lastName
        birthDate = updatingDto.birthDate
        gender = updatingDto.gender
        address = updatingDto.address
        email = updatingDto.email
        phoneNumber = updatingDto.phoneNumber
        updatedAt = LocalDateTime.now()
    }
}